### PR TITLE
Implement Profile

### PR DESCRIPTION
**Pull request by YehorPytomets**
### Rejections
We introduce rejections, that may be raised when handling the commands of the `ProfileAggregate`.
We redefine profile statuses to `PUBLISHED` and `NOT_PUBLISHED` and renamed some commands to reflect the profile status changes:
  - The profile cannot now be revoked because of ambiguity if the term _revoke_. But, it can now be marked as not published, given the published profile: `RevokeProfile` -> `MarkProfileAsNotPublished`;
  - The same is for _restoring from revoked_. Now we publish the profile instead: `RestoreProfileFromRevoked` -> `PublishProfile`.

### Profile Adding
- The `Owner` initiates adding the profile with the command `AddProfile` and receives back `ProfileAdded`;
- Restrict adding a profile with the already taken name:
Filter `ProfileNameChangingCommand`s, that provide already taken names, in `ProfileRepository` before the `ProfileAggregate` receives them;
- `ProfileAggregate`: 
  - restricts adding the same profile again -> `ProfileAlreadyAdded`;
  - adds a new profile and notifies views;
- `ProfileView`:
  - adds a profile view each time the profile is added.

### Profile Renaming
The `Owner` initiates renaming the profile with the command `RenameProfile` and receives back:
- `ProfileRenamed` event in case the profile was successfully renamed;
- `ProfileNotFound` if the profile doesn't exist at all;
- `InsufficientPermissionsForModifyingProfile` if the user pretended to be the profile owner;
- `CannotModifyRemovedProfile` if the profile is removed and was not restored back.
Or doesn't receive anything back if the new profile name is already taken by another profile. In this case, the command is not even delivered to the profile.

### Profile Removing
The profile `Owner` initiates removing the profile by sending the command `RemoveProfile` and receives back:
- `ProfileRemoved` event in case the profile was successfully removed;
- `ProfileNotFound` if the profile doesn't exist at all;
- `InsufficientPermissionsForModifyingProfile` if another actor pretended to be the profile owner;
- `CannotModifyRemovedProfile` if the profile has already been deleted.

### Profile Restoring
The profile is restored from removed after the profile `Owner` has sent the command to `RestoreRemovedProfile` and received back the `ProfileRestored` event.
The profile is rejected to be restored after sending the command because of one of the following reasons:
- another user has pretended to be the profile `Owner` -> `InsufficientPermissionsForModifyingProfile`;
- the profile doesn't exist at all -> `ProfileNotFound`;
- the profile hasn't been removed before -> `CannotRestoreNotRemovedProfile`.

### Profile Publishing
The profile is published after the profile `Owner` has sent the command to `PublishProfile` and received back the `ProfilePublished` event.
The profile is rejected to be published after sending the command because of one of the following reasons:
- another actor has pretended to be the profile `Owner` -> `InsufficientPermissionsForModifyingProfile`;
- the profile doesn't exist at all -> `ProfileNotFound`;
- the profile is removed -> `CannotModifyRemovedProfile`.

Publishing of an already published profile is allowed to cover the use case of introducing new changes to the profile.

### Marking the Profile as non-published
The profile `Owner` can mark the profile as not-published by sending the command `MarkProfileAsNonPublished`. 
This is helpful to hide the profile from being seen by everybody or before starting the work on a new version of the profile.
To mark the profile as non-published the owner must send the aforementioned command and receive back the `ProfileMarkedAsNonPublished` event.
The profile won't be marked in one of the following cases:
- the profile doesn't exist at all -> `ProfileNotFound`;
- another user has pretended to be the profile `Owner` -> `InsufficientPermissionsForModifyingProfile`;
- the profile is removed -> `CannotModifyRemovedProfile`;
- the profile must be published before marking it as not published -> `ProfileStatusMismatch`.

### Inviting Author
The profile `Owner` can invite an author to the profile by sending the command `InviteAuthor`. To become invited, the `ProfileAggregate` need to handle the command and emit the `AuthorInvited` event.
The author won't be invited in one of the following cases:
- the profile doesn't exist at all -> `ProfileNotFound`;
- another user has pretended to be the profile `Owner` -> `InsufficientPermissionsForModifyingProfile`;
- the profile is removed -> `CannotModifyRemovedProfile`;
- the author has already been invited or accepted the invitation -> `AuthorAlreadyAcceptedOrInvited`.

### Accepting Author Invitation
The profile `Author` can accept the invitation to the profile by sending the command `AcceptAuthorInvitation`. To accept the author, the `ProfileAggregate` need to handle the command and emit the `InvitationAcceptedByAuthor` event.
The author won't be able to accept the invitation in one of the following cases:
- the profile doesn't exist at all -> `ProfileNotFound`;
- the profile is removed -> `CannotModifyRemovedProfile`;
- the author hasn't been invited to the profile -> `AuthorMustBeInvited`.

### Rejecting Author Invitation
The profile `Author` can reject the invitation to the profile by sending the command `RejectAuthorInvitation`. To reject the invitation, the `ProfileAggregate` need to handle the command and emit the `InvitationRejectedByAuthor` event.
The author won't be able to reject the invitation in one of the following cases:
- the profile doesn't exist at all -> `ProfileNotFound`;
- the profile is removed -> `CannotModifyRemovedProfile`;
- the author hasn't been invited to the profile -> `AuthorMustBeInvited`.

### Removing Author
The profile `Owner` can remove an author from the profile by sending the command `RemoveAuthor`. To become removed, the `ProfileAggregate` need to handle the command and emit the `AuthorRemoved` event.
The author won't be removed in one of the following cases:
- the profile doesn't exist at all -> `ProfileNotFound`;
- another user has pretended to be the profile `Owner` -> `InsufficientPermissionsForModifyingProfile`;
- the profile is removed -> `CannotModifyRemovedProfile`;
- the author has been neither invited nor accepted the invitation -> `AuthorNotFound`.

### Rename: 
`QualificationProfile` -> `Profile`;
`ProfileNameAwareCommand` -> `ProfileNameChangingCommand`;
`QualificationProfileView` -> `ProfileView`;
`ProfileRestoredFromRemoved` -> `ProfileRestored`;
`QualificationProfileStatus` -> `ProfileStatus`.

### Update: 
- Spine `1.6.6` -> `1.6.8`;
- Some minor config changes.

### CODE REVIEW

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java [hunk 1]
```diff
@@ -26,81 +26,119 @@
...
+        builder().addAuthors(author);
     }
 
     @Subscribe
     void on(InvitationAcceptedByAuthor e) {
-        // TODO [Yehor Pytomets, 24.11.2020]: Subscribe on the 'InvitationAcceptedByAuthor' event.
+        List<ProfileAuthor> authors = state()
+                .getAuthorsList()
+                .stream()
+                .map(author -> author.getId()
```
Comment (armiol) [inline qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java:104]: Please extract the complex expression into a method.

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java [hunk 2]
```diff
@@ -26,81 +26,119 @@
...
+        builder().addAllAuthors(authors);
     }
 
     @Subscribe
     void on(InvitationRejectedByAuthor e) {
-        // TODO [Yehor Pytomets, 24.11.2020]: Subscribe on the 'InvitationRejectedByAuthor' event.
+        List<ProfileAuthor> authors = state()
+                .getAuthorsList()
+                .stream()
+                .filter(a -> !(a.getId()
```
Comment (armiol) [inline qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java:120]: Same as above.
You should extract into a method named in such a way, so that `.filter(...)` expression would read in English.

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java [hunk 3]
```diff
@@ -26,81 +26,119 @@
...
     @Subscribe
     void on(AuthorRemoved e) {
-        // TODO [Yehor Pytomets, 24.11.2020]: Subscribe on the 'AuthorRemoved' event.
+        List<ProfileAuthor> authors = state()
+                .getAuthorsList()
+                .stream()
+                .filter(author -> !author.getId()
+                                         .equals(e.getAuthor()))
+                .collect(toList());
+        builder().clearAuthors();
```
Comment (armiol) [inline qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java:136]: That's like the third time this code snippet is here. Please think of extracting the whole bit into something less repeatable.

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileRepository.java [hunk 4]
```diff
@@ -57,47 +54,45 @@
...
     }
 
     /**
-     * Obtains an identifier of the profile with a name that is provided by the command.
-     *
-     * <p>If no such profile found, then the {@link ProfileNameAwareCommand#getProfile() default}
-     * identifier is provided.
-     *
-     * <p>This helps to figure out and avoid profiles with duplicate names.
+     * Filters a {@code ProfileNameChangingCommand} if it provides a profile name,
```
Comment (armiol) [inline qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileRepository.java:61]: What do you mean by "filters"? Please rephrase.

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileRepository.java [hunk 5]
```diff
@@ -57,47 +54,45 @@
...
+    @SuppressWarnings("CheckReturnValue")
+    private ProfileId filter(ProfileNameChangingCommand cmd) {
 
-        ProfileId originalProfile = command.getProfile();
-        ProfileName name = command.getName();
+        ProfileName name = cmd.getName();
 
-        Optional<ProfileAggregate> mayBeProfile = findProfileByName(name);
+        if (hasProfileWithName(name)) {
+            newIllegalStateException("Cannot add a profile with the name '%s'. " +
```
Comment (armiol) [inline qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileRepository.java:75]: Is this about "adding" only? I thought we were handling a more general case here.

Please adjust the text.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/events.proto [hunk 6]
```diff
@@ -80,36 +80,25 @@ message ProfileRemoved {
...
-  ProfileId profile = 1 [(required) = true];
-}
-
-// A profile was restored from revoked by its owner.
-//
-// The state of the published profile should now be visible again for the users,
-// that are not its authors or the owner.
-//
-message ProfileRestoredFromRevoked {
+message ProfileMarkedAsNonPublished {
```
Comment (armiol) [inline qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/events.proto:101]: `UnpublishProfile`?

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/profile.proto [hunk 7]
```diff
@@ -33,58 +33,43 @@ option java_outer_classname = "ProfileProto";
...
+// the feature of design patterns is an important knowledge for a Java
+// developer. Whereas this may not be necessary for a PHP developer.
 //
-// These features may differ in each domain.
-// Some may be appropriate to one domain, while others may not.
-// For example, knowledge of design patterns is an integral part a software systems development and
-// API design.
-// However, this knowledge is not necessary for a machine learning engineer.
+// The same feature names can be used in different profiles.
+// For example, the Flyweight pattern can be both in Java and .NET
```
Comment (armiol) [inline qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/profile.proto:49]: Please quote the name.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto [hunk 8]
```diff
@@ -22,52 +22,97 @@ syntax = "proto3";
...
+message AuthorAlreadyAcceptedOrInvited {
+
+  // The profile identifier.
+  ProfileId profile = 1[(required) = true];
+
+  // The author identifier.
+  spine.core.UserId author = 2 [(required) = true];
+
+  // The author status.
+  ProfileAuthor.ProfileAuthorStatus status = 3 [(required) = true];
```
Comment (armiol) [inline qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto:86]: See? This is the one which must be shortened.

I also think that it might not be `Status`. I am not sure we speak of authors in terms of their statuses at all. It's an invitation which has a status. 

Please rethink and let's discuss.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto [hunk 9]
```diff
@@ -22,52 +22,97 @@ syntax = "proto3";
...
+}
+
+// The author must be invited to handle the command successfully.
+message AuthorMustBeInvited {
```
Comment (armiol) [inline qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto:90]: Did you mean `AuthorNotInvited` or `AuthorHasNoAccess`? We should speak stating the fact.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto [hunk 10]
```diff
@@ -22,52 +22,97 @@ syntax = "proto3";
...
+message AuthorNotFound {
+
+  // The profile identifier.
+  ProfileId profile = 1 [(required) = true];
+
+  // The author, that was not found.
+  spine.core.UserId author = 2 [(required) = true];
+}
+
+// Cannot handle a command, when the user doesn't have
```
Comment (armiol) [inline qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto:109]: Please rephrase so that's more like a description.

```
The rejection telling about insufficient permissions to modify some profile.

<more description goes into the following paragraph here>
```

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto [hunk 11]
```diff
@@ -44,73 +45,54 @@ message ProfileName {
...
+// An author is a user, responsible for filling the profile with features.
+//
+// The author adds drafts, modifies and prepares them for reviews.
+//
+message ProfileAuthor {
+
+  // The author identifier.
+  spine.core.UserId id = 1 [(required) = true];
+
+  // The author status.
```
Comment (armiol) [inline qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto:79]: Same as in my comment above. This field should be both named and described differently.

Try telling about the scenarios related to this field, to yourself. And you'll eventually find a proper term.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto [hunk 12]
```diff
@@ -44,73 +45,54 @@ message ProfileName {
...
+  ProfileAuthorStatus status = 2 [(required) = true];
+
+  // The profile author status in regard to the profile.
+  enum ProfileAuthorStatus {
```
Comment (armiol) [inline qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto:83]: See above on this matter. `ProfileAuthor.ProfileAuthorStatus` looks weird, at the very least :)

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/views.proto [hunk 13]
```diff
@@ -34,21 +34,24 @@ option java_outer_classname = "ViewsProto";
 option java_multiple_files = true;
 option java_generate_equals_and_hash = true;
 
-// The view of a single qualification profile.
+// A view of a qualification profile.
```
Comment (armiol) [inline qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/views.proto:37]: A view "on" a profile?

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/EntityTest.java [hunk 14]
```diff
@@ -80,6 +80,16 @@ protected void assertEntityStateMatches(Predicate<S> predicate) {
         assertTrue(matches);
     }
 
+    /**
+     * Asserts that the entity state matches another entity state.
+     *
+     * @param state
+     *         state to compare
+     */
+    protected void assertEntityStateMatches(S state) {
```
Comment (armiol) [inline qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/EntityTest.java:89]: Not sure what you tried to shorten by introducing this method.

In any case, it seems `final` to me.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Commands.java [hunk 15]
```diff
@@ -25,34 +25,101 @@
...
+
+    public static RejectAuthorInvitation rejectInvitation(Profile profile) {
+        return RejectAuthorInvitation
+                .newBuilder()
+                .setProfile(profile.getId())
+                .build();
+    }
+
+    public static RemoveAuthor removeAuthor(Profile profile, ProfileAuthor author) {
+        return RemoveAuthor.newBuilder()
```
Comment (armiol) [inline qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Commands.java:120]: Let's reformat this piece accordingly.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Events.java [hunk 16]
```diff
@@ -25,41 +25,105 @@
...
+    invitationRejected(Profile profile, ProfileAuthor author) {
+        return InvitationRejectedByAuthor
+                .newBuilder()
+                .setProfile(profile.getId())
+                .setAuthor(author.getId())
+                .vBuild();
+    }
+
+    public static AuthorRemoved authorRemoved(Profile profile, ProfileAuthor author) {
+        return AuthorRemoved.newBuilder()
```
Comment (armiol) [inline qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Events.java:124]: Same as previously.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Profiles.java [hunk 17]
```diff
@@ -46,81 +43,31 @@
...
 
-    public static QualificationProfile rename(QualificationProfile profile) {
-        return rename(profile, randomString());
-    }
-
-    public static QualificationProfile rename(QualificationProfile profile, String name) {
-        return profile
-                .toBuilder()
-                .setName(createProfileName(name))
+    public static UserId author(String id) {
```
Comment (armiol) [inline qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Profiles.java:67]: Please take a look at `GivenUserId` utility in Spine.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Rejections.java [hunk 18]
```diff
@@ -42,34 +50,75 @@
...
+    insufficientPermissions(Profile profile, UserId fakeOwner) {
+        return InsufficientPermissionsForModifyingProfile
+                .newBuilder()
+                .setProfile(profile.getId())
+                .setUser(fakeOwner)
+                .build();
+    }
+
+    public static CannotRestoreNotRemovedProfile cannotRestoreProfile(Profile profile) {
+        return CannotRestoreNotRemovedProfile
```
Comment (armiol) [inline qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Rejections.java:84]: It's also "non" here, but I would shorten this to `CannotRestoreProfile`, for simplicity. At this moment, we don't really care to distinguish the rejections at this level of detail.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Rejections.java [hunk 19]
```diff
@@ -42,34 +50,75 @@
...
+    public static AuthorMustBeInvited authorMustBeInvited(Profile profile, ProfileAuthor author) {
+        return AuthorMustBeInvited
+                .newBuilder()
+                .setProfile(profile.getId())
+                .setAuthor(author.getId())
+                .vBuild();
+    }
+
+    public static AuthorNotFound authorNotFound(Profile profile, ProfileAuthor author) {
+        return AuthorNotFound.newBuilder()
```
Comment (armiol) [inline qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Rejections.java:119]: It looks like every last method in these "given" classes is lacking some formatting.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/profiles/profile/AcceptAuthorInvitationTest.java [hunk 20]
```diff
@@ -0,0 +1,159 @@
...
+import static com.teamdev.qualifications.server.given.profiles.profile.Events.invitationAccepted;
+import static com.teamdev.qualifications.server.given.profiles.profile.ProfileViews.viewOf;
+import static com.teamdev.qualifications.server.given.profiles.profile.Profiles.author;
+import static com.teamdev.qualifications.server.given.profiles.profile.Profiles.name;
+import static com.teamdev.qualifications.server.given.profiles.profile.Profiles.profileId;
+import static com.teamdev.qualifications.server.given.profiles.profile.Rejections.authorMustBeInvited;
+import static com.teamdev.qualifications.server.given.profiles.profile.Rejections.cannotModifyRemovedProfile;
+import static com.teamdev.qualifications.server.given.profiles.profile.Rejections.profileNotFound;
+
+@DisplayName("'AcceptAuthorInvitation' command should")
```
Comment (armiol) [inline qualification-profiles/src/test/java/com/teamdev/qualifications/server/profiles/profile/AcceptAuthorInvitationTest.java:30]: To keep the code uniform with the framework and other projects on top of Spine, could you please change the single quote symbol (') you are using to a backtick (`) that we use in display-names?

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/profiles/profile/AddProfileTest.java [hunk 21]
```diff
@@ -0,0 +1,161 @@
...
+    void addProfileView() {
+        profilesContext()
+                .receivesCommand(add(profile()))
+                .assertState(profileId, ProfileView.class)
+                .comparingExpectedFieldsOnly()
+                .isEqualTo(viewOf(profile()));
+    }
+
+    @Test
+    @DisplayName("reject adding the already added profile")
```
Comment (armiol) [inline qualification-profiles/src/test/java/com/teamdev/qualifications/server/profiles/profile/AddProfileTest.java:84]: Please have a look at the start of the sentence. It tells that "a command should"; but here you say that "a command should reject ...".

Please review all such cases.

**Review by armiol — CHANGES_REQUESTED** [review]
@YehorPytomets please see my comments so far.

I guess, a lot of the test code is partly copied through the codebase. So it has all the same issues. Please incorporate my comments into the changes to similar portions of code.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto (hunk 8 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto:86]: Shorten the following:
`ProfileAuthor` -> `Author`;
`Author.ProfileAuthorStatus` -> `Author.InvitationStatus`;
`Author.InvitationStatus.INVITATION_ACCEPTED` -> `Author.InvitationStatus.ACCEPTED`.

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java (hunk 1 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java:104]: Extracted to the method with the following signature: 
`Author acceptInvitationIfAuthorsMatch(Author author, UserId acceptedAuthor)`

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java (hunk 2 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java:120]: Extract to the method with the following signature:
`boolean isAuthorInvited(Author profileAuthor, UserId userToCheck)`

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java (hunk 3 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java:136]: Extract to the method with the following signature:
`boolean checkAuthorsMatch(Author formerAuthor, UserId latterAuthor)`

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java (hunk 1 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java:104]: According to the discussion, rename to `Author acceptInvitationIfAuthorsSame(Author author, UserId acceptedAuthor)`

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java (hunk 3 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileProjection.java:136]: According to the discussion, rename to `boolean isSame(Author author, UserId user)`

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileRepository.java (hunk 4 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileRepository.java:61]: The intention of the method is to allow handling a `ProfileNameChangingCommand` if it provides the profile name, that is not taken by another profile.

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileRepository.java (hunk 5 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileRepository.java:75]: Adjusted to `Cannot handle the command with profile name '%s'. The name is already taken by another profile.`

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/events.proto (hunk 6 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/events.proto:101]: The Cambridge dictionary and the IDEA built-in dictionary doesn't recognize the word _unpublish_.
But the Oxford dictionary does.
Assuming that there is at least one reliable single source, the command `MarkProfileAsNonPublished` is renamed to `UnpublishProfile`. Related code is renamed as well.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/profile.proto (hunk 7 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/profile.proto:49]: Done

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto (hunk 9 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto:90]: `AuthorNotInvited`.
Renamed accordingly.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto (hunk 10 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/rejections.proto:109]: Renamed to `The user doesn't have sufficient permissions to modify the profile.`

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto (hunk 11 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto:79]: Renamed to
```
// The invitation status.
InvitationStatus status = 2 [(required) = true];
```
It sounds better when telling that the _invitation_ is either accepted or rejected, but not the profile author.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto (hunk 12 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto:83]: The `Author.InvitationStatus` looks better to me.

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/views.proto (hunk 13 above)
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/views.proto:37]: Yes, "view of" is another way to express the "one's understanding of smth". 
Renamed to "view on".

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Commands.java (hunk 15 above)
Comment (YehorPytomets) [reply qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Commands.java:120]: Done.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Events.java (hunk 16 above)
Comment (YehorPytomets) [reply qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Events.java:124]: Done

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Profiles.java (hunk 17 above)
Comment (YehorPytomets) [reply qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Profiles.java:67]: Done

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/EntityTest.java (hunk 14 above)
Comment (YehorPytomets) [reply qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/EntityTest.java:89]: Declared as `final`.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Rejections.java (hunk 19 above)
Comment (YehorPytomets) [reply qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Rejections.java:119]: Done

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Rejections.java (hunk 18 above)
Comment (YehorPytomets) [reply qualification-profiles/src/test/java/com/teamdev/qualifications/server/given/profiles/profile/Rejections.java:84]: Renamed to `CannotRestoreProfile`.
The same for the command: `RestoreRemovedProfile` -> `RestoreProfile`.
The event has already been renamed to `ProfileRestored`.

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/profiles/profile/AcceptAuthorInvitationTest.java (hunk 20 above)
Comment (YehorPytomets) [reply qualification-profiles/src/test/java/com/teamdev/qualifications/server/profiles/profile/AcceptAuthorInvitationTest.java:30]: Yes, fixed

File: qualification-profiles/src/test/java/com/teamdev/qualifications/server/profiles/profile/AddProfileTest.java (hunk 21 above)
Comment (YehorPytomets) [reply qualification-profiles/src/test/java/com/teamdev/qualifications/server/profiles/profile/AddProfileTest.java]: Done

**Comment by YehorPytomets** [comment]
@armiol , your comments have been taken into account. The PR is ready for review.

**Comment by YehorPytomets** [comment]
@armiol, the rejections were refactored according to the discussion. The PR is ready for review.

**Review by armiol — APPROVED** [review]
@YehorPytomets LGTM for the most part. Please address my comments and feel free to merge.

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileAggregate.java [hunk 22]
```diff
@@ -45,143 +46,509 @@
...
     private void event(ProfilePublished e) {
-        // TODO [Yehor Pytomets, 24.11.2020]: Apply the 'ProfilePublished' event.
+        builder().setStatus(PUBLISHED);
+    }
+
+    /**
+     * Accepts the author invitation if the same author is provided
+     *
+     * @param author
+     *         maybe accepted author
```
[the remark this answers is not in the document]
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileAggregate.java:349]: Done.

File: qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileAggregate.java [hunk 23]
```diff
@@ -45,143 +46,509 @@
...
+     * @param acceptedAuthor
+     *         the identifier of the author, that accepted the invitation
+     * @return the author with accepted invitation,
+     *         or the author without accepted invitation otherwise
+     */
+    private Author acceptInvitationIfAuthorSame(Author author, UserId acceptedAuthor) {
+        return isSame(author, acceptedAuthor) ?
+               acceptInvitation(author) :
```
[the remark this answers is not in the document]
Comment (YehorPytomets) [reply qualification-profiles/src/main/java/com/teamdev/qualifications/server/profiles/profile/ProfileAggregate.java:357]: Done

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/profile.proto [hunk 24]
```diff
@@ -33,58 +33,43 @@ option java_outer_classname = "ProfileProto";
...
 
-  // The tech expert of the profile.
-  spine.core.UserId tech_expert = 4;
+  // The profile lifecycle status.
+  ProfileStatus status = 4 [(required) = true];
 
-  // The lifecycle status of the profile.
-  QualificationProfileStatus status = 5 [(required) = true];
+  // The profile authors.
+  repeated Author authors = 5;
```
[the remark this answers is not in the document]
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/profile.proto:72]: Done

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto [hunk 25]
```diff
@@ -44,73 +45,55 @@ message ProfileName {
...
+    // The undefined value marker.
+    // Signifies that status is unset.
+    //
+    IS_UNDEFINED = 0;
+
+    // The owner has sent the user an invitation to become the profile author,
+    // but they haven't accepted or rejected it yet.
+    RECEIVED = 1;
+
+    // The user has accepted the invitation and become the profile author.
```
[the remark this answers is not in the document]
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto:94]: Done

File: qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto [hunk 26]
```diff
@@ -44,73 +45,55 @@ message ProfileName {
...
+  spine.core.UserId id = 1 [(required) = true];
+
+  // The author invitation.
+  Invitation invitation = 2 [(required) = true];
+
+  // The status of the author invitation in regard to the profile.
+  enum Invitation {
+
+    // The undefined value marker.
+    // Signifies that status is unset.
```
[the remark this answers is not in the document]
Comment (YehorPytomets) [reply qualification-profiles/src/main/proto/com/teamdev/qualifications/profiles/profile/values.proto:86]: Done.
